unit/java_bytecode/java_object_factory/gen_nondet_string_init.cpp

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************\
+
+ Module: Java string library preprocess.
+         Test for converting an expression to a string expression.
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <util/expr.h>
+#include <util/std_code.h>
+#include <util/namespace.h>
+#include <java_bytecode/java_object_factory.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/java_root_class.h>
+#include <langapi/language_util.h>
+#include <langapi/mode.h>
+#include <iostream>
+
+SCENARIO("Generate string object",
+  "[core][java_bytecode][java_object_factor][gen_nondet_string_init]")
+{
+  GIVEN("an expression, a location, and a symbol table")
+  {
+    source_locationt loc;
+    symbol_tablet symbol_table;
+    register_language(new_java_bytecode_language);
+
+    // Add java.lang.Object to symbol table
+    symbolt jlo_sym;
+    jlo_sym.name="java::java.lang.Object";
+    jlo_sym.type=struct_typet();
+    jlo_sym.is_type=true;
+    java_root_class(jlo_sym);
+    bool failed=symbol_table.add(jlo_sym);
+    CHECK_RETURN(!failed);
+
+    // Add java.lang.String to symbol table
+    java_string_library_preprocesst preprocess;
+    preprocess.add_string_type("java.lang.String", symbol_table);
+    namespacet ns(symbol_table);
+
+    // Declare a String named arg
+    symbol_typet java_string_type("java::java.lang.String");
+    symbol_exprt expr("arg", java_string_type);
+
+    WHEN("Initialisation code for a string is generated")
+    {
+      const codet code=
+        initialize_nondet_string_struct(expr, 20, loc, symbol_table);
+
+      THEN("Code is produced")
+      {
+        std::vector<std::string> code_string;
+
+        const std::regex spaces("\\s+");
+        const std::regex numbers("\\$[0-9]*");
+        for(auto op : code.operands())
+        {
+          const std::string line=from_expr(ns, "", op);
```

Comment:
I'm not going to block on this, but I think in general we should avoid writing unit tests that rely on the behaviour of `from_expr` as it makes the tests quite brittle. Instead, we've been trying to introduce helper functions to make writing code that actually checks, `exprt`/`typet`/`codet` conform to a requirement. 

E.g, for below you'd do something like
```C++
require_declaration(code.operands[0], "tmp_object_factory", java_int_type());
```
To verify the first instruction is a declaration of a variable called `tmp_object_factory` (if that is important) of type `int`

Reply:
that would be nice